digdag-core/src/main/java/io/digdag/core/database/BasicDatabaseStoreManager.java

```diff
@@ -274,42 +283,105 @@ public int getIndex()
+    {
+        Handle nested = currentTransaction.get();
+        if (nested != null) {
+            return action.call(nested, nested.attach(daoIface));
         }
-        catch (InterruptedException ex) {
-            throw Throwables.propagate(ex);
+        else {
+            try {
+                return transactionRetryExecutor.runInterruptible(() -> {
+                    try (Handle handle = dbi.open()) {
+                        currentTransaction.set(handle);
+                        handle.begin();
+                        T retval;
+                        try {
+                            retval = action.call(handle, handle.attach(daoIface));
+                        }
+                        catch (Exception ex) {
+                            try {
+                                handle.rollback();
+                            }
+                            catch (Exception rollback) {
+                                ex.addSuppressed(rollback);
+                            }
+                            Throwables.propagateIfInstanceOf(ex, exClass1);
+                            Throwables.propagateIfInstanceOf(ex, exClass2);
+                            Throwables.propagateIfPossible(ex);
+                            throw new TransactionFailedException(
+                                    "Transaction failed due to exception being thrown " +
+                                    "from within the callback. See cause " +
+                                    "for the original exception.", ex);
+                        }
+                        handle.commit();
+                        return retval;
+                    }
+                    finally {
+                        currentTransaction.set(null);
+                    }
+                });
+            }
+            catch (RetryGiveupException ex) {
+                Throwable innerException = ex.getCause();
+                Throwables.propagateIfInstanceOf(innerException, exClass1);
+                Throwables.propagateIfInstanceOf(innerException, exClass2);
+                throw Throwables.propagate(innerException);
+            }
+            catch (InterruptedException ex) {
+                throw Throwables.propagate(ex);
+            }
         }
     }
 
     @SuppressWarnings("unchecked")
-    public <T, E1 extends Exception, E2 extends Exception> T transaction(TransactionActionWithExceptions<T, D, E1, E2> action, Class<E1> exClass1, Class<E2> exClass2) throws E1, E2
+    public <T, E1 extends Exception, E2 extends Exception> T transactionWithLocalRetry(RetryableTransactionActionWithExceptions<T, D, E1, E2> action, Class<E1> exClass1, Class<E2> exClass2) throws E1, E2
     {
+        // TODO
+        // Here should throw error if currentTransaction.get() != null to avoid race condition under high load.
```

Comment:
I also tried to do that but I couldn't because DatabaseSecretControlStore already depends on it. Well, actually DatabaseSecretControlStore.setProjectSecret seems broken...(it throws conflict exception if 2 threads run the method concurrently).